src/ucp/dt/datatype_iter.c

```diff
@@ -265,11 +265,6 @@ size_t ucp_datatype_iter_iov_next_iov(const ucp_datatype_iter_t *dt_iter,
     ucs_assertv(length <= max_length, "length=%zu max_length=%zu", length,
                 max_length);
 
-    /* Check that if not reached the end, packed data is not empty */
-    ucs_assertv((dt_iter->offset == dt_iter->length) || (length > 0),
```

Comment:
Is that true that all middle fragments still should always send some payload? Maybe somehow save this check for middle segments only (e.g. check offset)?

Reply:
Yes, it's true
To have this assert I would need to propagate request (or at least request flags/boolean) to this function and to all the functions calling it (`ucp_datatype_iter_next_iov` refs in 7 other files). If it's really important to have this check, we can do that.
@yosefe - since you added this assertion, do we need to keep it even at the cost of adding one more param to this function?

Reply:
I don't think that it is necessary to add request argument here. We can pass only information that is needed and not to this function but to `ucp_datatype_iter_next_iov`. Also maybe we can do that even without passing request at all by relaxing this assertion (E.g. check that only if offset > 0) WDYT?

Reply:
Ok, I add relaxed check, still maybe useful